lang/cs/Org.Apache.REEF.Network/Elastic/Task/Workflow.cs

```diff
@@ -0,0 +1,298 @@
+        {
+            lock (_lock)
+            {
+                if (!_disposed)
+                {
+                    if (_operators != null)
+                    {
+                        // Clean dispose, check that the computation is completed
+                        if (!_failed)
+                        {
+                            foreach (var op in _operators)
+                            {
+                                if (op != null)
+                                {
+                                    op.WaitCompletionBeforeDisposing();
+                                }
+                            }
+                        }
+
+                        foreach (var op in _operators)
+                        {
+                            if (op != null)
+                            {
+                                var disposableOperator = op as IDisposable;
+
+                                disposableOperator.Dispose();
+                            }
+                        }
+                    }
+
+                    _disposed = true;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Add an elastic operator to the workflow.
+        /// </summary>
+        /// <param name="op"></param>
+        internal void Add(IElasticOperator op)
+        {
+            op.CancellationSource = _cancellationSource.Source;
+
+            _operators.Add(op);
+
+            if (_iteratorsPosition.Count > 0)
+            {
+                var iterPos = _iteratorsPosition.Last();
+                var iterator = _operators[iterPos] as IElasticIterator;
+
+                op.IteratorReference = iterator;
+                iterator.RegisterActionOnTaskRescheduled(op.OnTaskRescheduled);
+            }
+
+            if (op.OperatorName == Constants.Iterate)
+            {
+                _iteratorsPosition.Add(_operators.Count - 1);
+            }
+        }
+
+        /// <summary>
+        /// Initializes the communication group.
+        /// Computation blocks until all required tasks are registered in the group.
+        /// </summary>
+        /// <param name="cancellationSource">The signal to cancel the operation</param>
+        internal void WaitForTaskRegistration(CancellationTokenSource cancellationSource = null)
+        {
+            try
+            {
+                foreach (var op in _operators)
+                {
+                    op.WaitForTaskRegistration(cancellationSource);
+                }
+            }
+            catch (OperationCanceledException e)
+            {
+                throw e;
+            }
+        }
+
+        /// <summary>
+        /// Reset the position tracker for all operators in the workflow.
+        /// </summary>
+        private void ResetOperatorPositions()
+        {
+            for (int pos = _position; pos < _operators.Count; pos++)
+            {
+                _operators[pos].ResetPosition();
+            }
```

Comment:
why not just
```C#
foreach (var op in _operators)
{
    op.ResetPosition();
}
```

Reply:
This is tricky. When you have iterators in the pipeline you only reset each time within the context of an iteration. The for loop is used because the position tells where to actually start to reset.

Reply:
Yes. `ResetOperatorPositions` is private and only used inside `MoveNext` where the first operation is _position++.